### ISSUE TITLE
Inconsistensy of container descriptor default value type for nominalizable types.

### ISSUE THREAD

**Issue by vrurg** [perl6/problem-solving#3]
May I propose an easy problem to start with?

The type of container descriptor `default` attribute depends on what particular nominalizable type was used to declare a variable/attribute. For a definite it would be the base type, for a subset – the subset itself:

```
my Int:D $a = 0; 
say $a.VAR.default.^name; # Int
my Int $a where {True};
say $a.VAR.default.^name, " of ", $a.VAR.default.HOW.^name; # <anon> of Perl6::Metamodel::SubsetHOW
```

I would suggest that in both cases the implicit default value must be the base type, i.e. `Int` in the example above. More precisely, the declaration type of the variable must be nominalized and the resulting type object is to be used for default. It would mean that:

```
my Int:D $a where {True} = 0;
say $a.VAR.default.^name; # Int
```

**Comment by vrurg** [comment]
Initial proposal:

While working on a fix for [https://github.com/rakudo/rakudo/issues/2166](Rakudo#2166) I came across a code in `World`'s `container_type_info` where container descriptor is initialized with a result of `maybe-definite-how-base` method. Generalizing it with nominalization would produce the desired result.

I tried to get this done. The change contradicts current standards and breaks roast tests where the default for a subset is expected to be the subset itself.

**Comment by vrurg** [comment]
Personal consideration. I think this is the right way to have things. Basically, it's been already done right for the smileys. When we declare a variable/attribute we do it with a _type_ and apply some _constraints_.  For example:

```
subset S of Str where *.chars > 2;
my S $a is default("the answer is 42");
```

then descriptor's default type will be `Str`, not `S`. On top of it, if some code relies on the default then it would need to check for the metaclass archetype. Simply compare:

```
if $a.VAR.default.WHAT === Int {
    ...
}
```

to

```
my \default = $a.VAR.default.HOW.archetypes.nominalizable ?? $a.VAR.default.^nominalize !! $a.VAR.default;
if default.WHAT === Int {
    ...
}
```

And back to what I said in the beginning, since a definite could be declared as a subset, it would be a good point for that subset to behave similarly to the definite (except for the requirement to have an initializer).

**Comment by AlexDaniel** [comment]
Thank you for this ticket! Definitely a test drive opportunity for this repo! As it affects roast, attaching a `language` label (of course it is also Rakudo-related, but that'd be dispatched to the same person anyway).

**Comment by jnthn** [comment]
In this case:

```
my Int:D $a = 0; 
say $a.VAR.default.^name; # Int
```

The value of `default` is not used; if we were to write:

```
my Int:D $a;
```

Then it will not compile. So I'm not sure that the value of `default` tells us a great deal here; it's probably a side-effect of the rather more useful case of:

```
my Int:U $x;
```

Where `Int` being the default is both useful and observable.

> it would be a good point for that subset to behave similarly to the definite (except for the requirement to have an initializer)

I'd be more inclined to fully align this with the `:D` case and *require* the initializer, unless the type object forming the default `default` would match the subset constraint (this check would be done under `try` semantics, and an error in evaluating the `subset` type treated as a non-match). That would mean the user has one common rule to remember: if the type is constrained in such a way that it cannot hold a type object, then an initializer or an `is default` is required. As you suggest, the type object in question here would be that of the base type, not of the `subset` type.

So, I'm agreeing that the `default` be the base type, but also wishing to take the logical next step while we're changing this anyway.

It's a slight aside, but since `is default` was mentioned: I think we should generally prefer leading folks towards an initializer rather than `is default`, because it isn't surprising in behavior. By contrast, the expression given to `is default` is evaluated once, at compile time. That's sensible: a trait is applied at compile time. However, I suspect in many cases the thing to assign would be something other than a literal.

> The change contradicts current standards and breaks roast tests where the default for a subset is expected to be the subset itself.

Indeed; this would need to be done in 6.e already.

**Comment by vrurg** [comment]
> I'd be more inclined to fully align this with the :D case and require the initializer, unless the type object forming the default default would match the subset constraint

Funny, but that was my initial intention to go this way. But later I put higher priority on the backward compatibility of the fix, so rakudo/rakudo#2727 contains a fix which keeps current roast happy. I would keep it this way to make it possible for the PR to be applied to 6.d; and then I will produce a new PR for the upcoming 6.e which would introduce the more reasonable semantics. So for

```
subset S1 of Int where *.defined;
subset S2 of S1 where {True};
my S2 $foo = 1;
```

`$foo.VAR.default` will be `Int` and the code fill fail without the initializer.

As to the `is default` matters in general, I personally don't like it to to be strictly compile-time as this limits a range of great applications for the trait. For example:

```
sub foo ($a is raw) {
    $a = Nil;
}
sub bar ($def) {
    my $v is default($def);
    foo($v); # $v is now Mu
}
```

Whatever would be the reason for `foo` to reset a variable to its default value – it has the right to do so. The ability itself would spare a few lines and complexity of some code.

This issue is workarounded in my `AttrX::Mooish` with laziness and attribute clearers, but the approach is not applicable to lexicals.

Anyway, this goes all the wrong way now. `is default` is worth separate discussion but not under this topic, I think.

**Comment by vrurg** [comment]
Unfortunately, I'm currently stuck with rakudo/rakudo#2772 which prevents me from proceeding.

**Comment by vrurg** [comment]
I now have a patch on my hands in https://github.com/vrurg/rakudo/tree/problem-solving-3. Patches to roast are still to come.

What I have no idea about at this moment is how to handle the patches now. It doesn't seem to me like merging into the `master` is the right thing to do. But seemingly there is no 6e branch either.

**Comment by AlexDaniel** [comment]
> It doesn't seem to me like merging into the master is the right thing to do. But seemingly there is no 6e branch either.

IMO we shouldn't have a 6.e branch. Generally, changes you make to the compiler should only have effect if the code does `use v6.e`. That said, there are cases when it's ok to change things for all language versions, for example if that aspect was not spectested or if the test was unambiguously wrong.

**Comment by vrurg** [comment]
This is definitely not the latter case. The following:

```
subset S1 of Int:D where { True };
my S1 $s1;
subset S2 of Int where *.defined;
my S2 $s2;
```

Wouldn't be valid code anymore. Meanwhile, `6.e` isn't opened yet. Maybe it is time to add it into `gen-version.pl`? Or how otherwise the code could be tested?

**Comment by AlexDaniel** [comment]
> Meanwhile, 6.e isn't opened yet. Maybe it is time to add it into gen-version.pl?

Yes. https://github.com/rakudo/rakudo/issues/2783

See also https://github.com/rakudo/rakudo/issues/1289

**Comment by vrurg** [comment]
With regard to versioning, in this case we're lucky because it is enough to know what version the subset was compiled with. I don't think it's a good idea for a type object to behave differently under different `v6.X` pragmas. 

Another problem with versions is a high risk of spaghetti code. In either case this matter must be carefully considered and discussed in a dedicated thread.

And my task now is to decide how to effectively combine rakudo/rakudo#2727 for `6.d` with what must go into `6.e`.

**Comment by vrurg** [comment]
I finally can get back to this problem. And when it was said that the problem is _easy_ it was way too preliminary opinion...

Now, when I have `6.e` to experiment with another side of the problem has been revealed. rakudo/rakudo#2946 contains the changes I've done so far. The [problematic line](https://github.com/rakudo/rakudo/pull/2946/files#diff-6376f37e1e72677e1e378984eb5958edR80) is the core of the fix.

The problem #1 is that it breaks a lot of spectests at compilation stage. I didn't investigate this yet. But this breakage raised a question of mixing up code of different versions. For example:

```
unit module Mod6d;

subset S1 of Str is export where *.defined;

sub foo ($s) {
    $s ~~ $s
}

foo(S1); # Pass
```

```
use v6.e;
use Mod6d;

subset S2 of Int where *.defined;

foo(S2); # ??
S1 ~~ S1; # ??
S2 ~~ S2; # Fail
```

While it is clear that `S1` must match against itself within `Mod6d` while `S2` matching must fail under `v6.e` pragma – it is questionable how they must behave under another version? The options I see here:

1. `6.e` behavior takes over.

   Both `S1` and `S2` would always fail to match. This has the price of breaking many ecosystem modules.

1. A type follows the rules of the version it was created under.

   I.e. `S1` would always match while `S2` would always fail.

1. A type follows the rules of the calling code.

   This is currently impossible to implement. 

My own favor is for the first option and I wouldn't doubt going this way would only there be a way to provide a grace period to fix the modules. But even when modules are adapted for the new behavior the other problem will remain: how to provide bug fixes for the users of older releases?

**Comment by vrurg** [comment]
Out of my curiosity, I have implemented item 2 from the previous comment and it works as expected. I.e. in the example above `S1 ~~ S1` is _True_. Horrible, but spectests are seemingly happy about it...

What else has crossed my mind: if item 1 is chosen then what breaks are:

```
subset SS of Str where *.defined;

my SS $var;
``` 

and 

```
if SS ~~ SS {
    ...
}
```

The fixes are as simple as:

```
my SS $var = "default";
if SS === SS {
    ...
}
```

Which are the right ways for this anyway. Perhaps some code would expect `SS` to match both `Str` and itself, but I hardly see what real use this might have.

**Comment by vrurg** [comment]
While figuring out how to deal with `roast` I've got one more reason for following the first option of breaking the backward compatibility. Preserving it means that in the following case:

```
subset SS of Int where * > 10;
class Foo {
    has SS $.foo;
}
```

`$.foo`'s implicit default is set to `SS`. Whereas it was ok when `SS ~~ SS` is _true_, now it results in an attempt to match `SS` against `Int` by `Actions.nqp` `check_default_value` routine. As a result, the core tries to cast `SS` into `Real` and obviously fails.

Perhaps it's the  _all or nothing_ case...

**Comment by vrurg** [comment]
Added perl6/roast#547 with proposed changes to roast. The PR also demonstrates what kind of changes would be needed to conform the proposed `subset` logic.

**Comment by jnthn** [comment]
> This has the price of breaking many ecosystem modules.

We try really hard to avoid this outside of language version boundaries. The ecosystem is only a subset of the Perl 6 code in existence, so anything breaking many modules there will have consequences for other code. Back compat is no fun, but a reputation for knowingly regularly breaking stuff is even less fun.

> A type follows the rules of the version it was created under.

This at least means that one can't come across such a type in code that already exists.

**Comment by vrurg** [comment]
So, the choice would be to remember what version a subset was created with and follow the rules of that version? I'm not sure what effect would it has if such type is passed from one revision to another.

```
use v6.d;
unit module Mod-d;

subset SS of Int is export where * > 10;

sub foo is export {
    my SS $v;
    return $v;
}
```

```
use v6.e;
module Mod-e;
use Mod-d;

my $v = foo;

say $v.defined; # False

if $v ~~ SS {
   ... # How come we happened here?
} 
```

How could one guess that undefined `$v` would match `SS` without knowing that `Mod-d` is utilizing revision `d`? And what happens when `Mod-d` switches to `e` at some point?

There is a good reason to for breaking the backward compatibility in this case. The old behavior is a bug in its nature. I.e. `my SS $v;` must not be valid without an initializer. It is not something we break just for the sake of a change.

**Comment by vrurg** [comment]
Perhaps, by reading @jnthn word to word, I would add a fourth item to the list of options:

4. Do as suggested in my [Mar 1 comment](https://github.com/perl6/problem-solving/issues/3#issuecomment-468871849) and only nominalize the default value with `use v6.e;` without changing the smartmatch rules for subsets. This would keep full backward compatibility at run time but will contradict to [Jonathan's comment](https://github.com/perl6/problem-solving/issues/3#issuecomment-472648352).

**Comment by ugexe** [comment]
> This has the price of breaking many ecosystem modules.

Can you quantify this? You have some MOP type modules, so maybe you're overestimating how bad it would be.

**Comment by vrurg** [comment]
> > This has the price of breaking many ecosystem modules.
> 
> Can you quantify this? You have some MOP type modules, so maybe you're overestimating how bad it would be.

Lack of experience & hunting for build bugs... Will try Blin and see what it says.

**Comment by AlexDaniel** [comment]
@vrurg Blin currently requires a commit to master, so it won't help for experimental changes, not right now at least. It just happens that branches are not tracked by whateverable, so that needs to be fixed first…

**Comment by vrurg** [comment]
Thanks Alex, I thought it's me doing wrong. Will try installing by a list then.

Anyway, while installing Blin dependencies I came across HTTP::Request which is incompatible with the change. Again, cloned, added initializer to an attribute and it went on ok. All other modules were installed seamlessly.

**Comment by vrurg** [comment]
Bruteforce `zef install --install-to=./lib` by scanning `META.list` from ecosystem repo. The script managed to try 428 installations then something stuck. Not sure what exactly as I tried installing with 10 processes in parallel. But hope it's not relevant.

Of 428:

213 just failed for whatever reason (no deps, bad tests, etc.)
33 of them failed with due to the changed behavior.
12 of those are the modules actually affected by the change. 

The other 21 were dependent on these 12 affected. Therefore I can't tell if they're affected or not.  Let's say 2-3 would have the problem too.

Most of the 180 modules may be ignored because they fail anyway.

Extrapolating to all 866 modules from the ecosystem gives an estimate of 20-40 problematic modules. Not a few, but not that many as I was afraid of. For those I can personally iterate over and produce PRs. They're often would be as simple as changing something like:

```
has Int $.foo where 1 | 2;
```

to one of:

```
has Int $.foo where Int:U | 1 | 2;
has Int $.foo where 1 | 2 = 1;
```

depending on context. Because whatever the final decision is be, the original line looks like something deeply incorrect to me as it allows for undefined value where a strict constraint is applied.

**Comment by vrurg** [comment]
I have implemented the 'source language version defines subset semantics' logic and merged it. If this won't cause a big disaster then this ticket can be closed.

**Comment by vrurg** [comment]
Closing after merge of rakudo/rakudo#2946
